File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/openclaw/ToolCallModels.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw

import com.meta.wearable.dat.externalsampleapps.cameraaccess.conference.ConferenceSourceType
import org.json.JSONArray
import org.json.JSONObject

// Gemini Tool Call (parsed from server JSON)

data class GeminiFunctionCall(
    val id: String,
    val name: String,
    val args: Map<String, Any?>
)

data class GeminiToolCall(
    val functionCalls: List<GeminiFunctionCall>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCall? {
            val toolCall = json.optJSONObject("toolCall") ?: return null
            val calls = toolCall.optJSONArray("functionCalls") ?: return null
            val functionCalls = mutableListOf<GeminiFunctionCall>()
            for (i in 0 until calls.length()) {
                val call = calls.getJSONObject(i)
                val id = call.optString("id", "")
                val name = call.optString("name", "")
                if (id.isEmpty() || name.isEmpty()) continue
                val argsObj = call.optJSONObject("args")
                val args = mutableMapOf<String, Any?>()
                if (argsObj != null) {
                    for (key in argsObj.keys()) {
                        args[key] = argsObj.opt(key)
                    }
                }
                functionCalls.add(GeminiFunctionCall(id, name, args))
            }
            return if (functionCalls.isNotEmpty()) GeminiToolCall(functionCalls) else null
        }
    }
}

// Gemini Tool Call Cancellation

data class GeminiToolCallCancellation(
    val ids: List<String>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCallCancellation? {
            val cancellation = json.optJSONObject("toolCallCancellation") ?: return null
            val idsArray = cancellation.optJSONArray("ids") ?: return null
            val ids = mutableListOf<String>()
            for (i in 0 until idsArray.length()) {
                ids.add(idsArray.getString(i))
            }
            return if (ids.isNotEmpty()) GeminiToolCallCancellation(ids) else null
        }
    }
}

// Tool Result

sealed class ToolResult {
    data class Success(val result: String) : ToolResult()
    data class Failure(val error: String) : ToolResult()

    fun toJSON(): JSONObject = when (this) {
        is Success -> JSONObject().put("result", result)
        is Failure -> JSONObject().put("error", error)
    }
}

// Tool Call Status (for UI)

sealed class ToolCallStatus {
    data object Idle : ToolCallStatus()
    data class Executing(val name: String) : ToolCallStatus()
    data class Completed(val name: String) : ToolCallStatus()
    data class Failed(val name: String, val error: String) : ToolCallStatus()
    data class Cancelled(val name: String) : ToolCallStatus()

    val displayText: String
        get() = when (this) {
            is Idle -> ""
            is Executing -> "Running: $name..."
            is Completed -> "Done: $name"
            is Failed -> "Failed: $name - $error"
            is Cancelled -> "Cancelled: $name"
        }

    val isActive: Boolean
        get() = this is Executing
}

// OpenClaw Connection State

sealed class OpenClawConnectionState {
    data object NotConfigured : OpenClawConnectionState()
    data object Checking : OpenClawConnectionState()
    data object Connected : OpenClawConnectionState()
    data class Unreachable(val message: String) : OpenClawConnectionState()
}

// Tool Declarations (for Gemini setup message)

object ToolDeclarations {
    const val EXECUTE_NAME = "execute"
    const val EXTRACT_ENTITY_NAME = "extract_entity"

    data class ToolDeclaration(
        val name: String,
        val description: String,
        val parameters: Map<String, Any>,
        val behavior: String,
    )

    fun allDeclarations(conferenceModeEnabled: Boolean = false): List<ToolDeclaration> {
        return buildList {
            add(executeDeclaration())
            if (conferenceModeEnabled) {
                add(extractEntityDeclaration())
            }
        }
    }

    fun allDeclarationsJSON(conferenceModeEnabled: Boolean = false): JSONArray {
        return JSONArray().apply {
            allDeclarations(conferenceModeEnabled).forEach { declaration ->
                put(declaration.toJSON())
            }
        }
    }

    private fun executeDeclaration(): ToolDeclaration {
        return ToolDeclaration(
            name = EXECUTE_NAME,
            description = "Your only way to take action. You have no memory, storage, or ability to do anything on your own -- use this tool for everything: sending messages, searching the web, adding to lists, setting reminders, creating notes, research, drafts, scheduling, smart home control, app interactions, or any request that goes beyond answering a question. When in doubt, use this tool.",
            parameters = mapOf(
                "type" to "object",
                "properties" to mapOf(
                    "task" to mapOf(
                        "type" to "string",
                        "description" to "Clear, detailed description of what to do. Include all relevant context: names, content, platforms, quantities, etc.",
                    ),
                ),
                "required" to listOf("task"),
            ),
            behavior = "BLOCKING",
        )
    }

    private fun extractEntityDeclaration(): ToolDeclaration {
        return ToolDeclaration(
            name = EXTRACT_ENTITY_NAME,
            description = "Local-only conference mode extraction tool. Use it to silently report a detected badge, business card, booth sign, or slide without speaking.",
            parameters = mapOf(
                "type" to "object",
                "properties" to mapOf(
                    "name" to mapOf(
                        "type" to "string",
                        "description" to "Detected person or entity name.",
                    ),
                    "company" to mapOf(
                        "type" to "string",
                        "description" to "Detected company or organization name.",
                    ),
                    "role" to mapOf(
                        "type" to "string",
                        "description" to "Detected job title or role.",
                    ),
                    "source_type" to mapOf(
                        "type" to "string",
                        "enum" to ConferenceSourceType.entries.map { it.wireValue },
                        "description" to "Where the entity was detected.",
                    ),
                    "confidence" to mapOf(
                        "type" to "number",
                        "description" to "Confidence score from 0.0 to 1.0.",
                    ),
                    "observed_text" to mapOf(
                        "type" to "string",
                        "description" to "Optional raw snippet seen in the frame.",
                    ),
                ),
                "required" to listOf("name", "source_type", "confidence"),
            ),
            behavior = "NON_BLOCKING",
        )
    }

    private fun ToolDeclaration.toJSON(): JSONObject {
        return JSONObject().apply {
            put("name", name)
            put("description", description)
            put("parameters", parameters.toJSONObject())
            put("behavior", behavior)
        }
    }

    private fun Map<String, *>.toJSONObject(): JSONObject {
        return JSONObject().apply {
            entries.forEach { (key, value) ->
                put(key, value.toJsonValue())
            }
        }
    }

    private fun Any?.toJsonValue(): Any {
        return when (this) {
            null -> JSONObject.NULL
            is Map<*, *> -> {
                val mapped = entries.associate { (key, value) -> key.toString() to value }
                mapped.toJSONObject()
            }
            is List<*> -> JSONArray().apply {
                this@toJsonValue.forEach { item ->
                    put(item.toJsonValue())
                }
            }
            else -> this
        }
    }
}

File: samples/CameraAccess/CameraAccess/OpenClaw/ToolCallModels.swift
import Foundation

// MARK: - Gemini Tool Call (parsed from server JSON)

struct GeminiFunctionCall {
  let id: String
  let name: String
  let args: [String: Any]
}

struct GeminiToolCall {
  let functionCalls: [GeminiFunctionCall]

  init?(json: [String: Any]) {
    guard let toolCall = json["toolCall"] as? [String: Any],
          let calls = toolCall["functionCalls"] as? [[String: Any]] else {
      return nil
    }
    self.functionCalls = calls.compactMap { call in
      guard let id = call["id"] as? String,
            let name = call["name"] as? String else { return nil }
      let args = call["args"] as? [String: Any] ?? [:]
      return GeminiFunctionCall(id: id, name: name, args: args)
    }
  }
}

// MARK: - Gemini Tool Call Cancellation

struct GeminiToolCallCancellation {
  let ids: [String]

  init?(json: [String: Any]) {
    guard let cancellation = json["toolCallCancellation"] as? [String: Any],
          let ids = cancellation["ids"] as? [String] else {
      return nil
    }
    self.ids = ids
  }
}

// MARK: - Local glasses media requests

enum GlassesMediaKind: String, Equatable {
  case snapshot
  case video
}

struct GlassesMediaRequest: Equatable {
  let kind: GlassesMediaKind
  let requestedDurationSeconds: Int?

  init?(args: [String: Any]) {
    guard let rawKind = args["kind"] as? String,
          let kind = GlassesMediaKind(rawValue: rawKind) else {
      return nil
    }
    self.kind = kind
    if let duration = args["durationSeconds"] as? Int {
      self.requestedDurationSeconds = min(max(duration, 1), 30)
    } else if let duration = args["durationSeconds"] as? Double {
      self.requestedDurationSeconds = min(max(Int(duration.rounded()), 1), 30)
    } else {
      self.requestedDurationSeconds = nil
    }
  }
}

// MARK: - Tool Result

enum ToolResult {
  case success(String)
  case failure(String)

  static let maxResponseCharacters = 12_000

  private static func bounded(_ value: String) -> String {
    guard value.count > maxResponseCharacters else { return value }
    let marker = "\n\n[response truncated]"
    return String(value.prefix(maxResponseCharacters - marker.count)) + marker
  }

  var responseValue: [String: Any] {
    switch self {
    case .success(let result):
      return ["result": Self.bounded(result)]
    case .failure(let error):
      return ["error": Self.bounded(error)]
    }
  }
}

// MARK: - Tool Call Status (for UI)

enum ToolCallStatus: Equatable {
  case idle
  case executing(String)
  case completed(String)
  case failed(String, String)
  case cancelled(String)

  var displayText: String {
    switch self {
    case .idle: return ""
    case .executing(let name): return "Running: \(name)..."
    case .completed(let name): return "Done: \(name)"
    case .failed(let name, let err): return "Failed: \(name) - \(err)"
    case .cancelled(let name): return "Cancelled: \(name)"
    }
  }

  var isActive: Bool {
    if case .executing = self { return true }
    return false
  }
}

// MARK: - Tool Declarations (for Gemini setup message)

enum ToolDeclarations {

  static func allDeclarations(
    namedRoutingEnabled: Bool = false,
    registry: NamedHarnessRegistry = .standard()
  ) -> [[String: Any]] {
    var declarations: [[String: Any]]
    if namedRoutingEnabled {
      declarations = [routeHarness(registry: registry)]
    } else {
      declarations = [execute]
    }
    if GlassesSessionFeaturePolicy.voiceSnapshotEnabled {
      declarations.append(captureMedia)
    }
    return declarations
  }

  static let execute: [String: Any] = [
    "name": "execute",
    "description": "Your mandatory connection to OpenClaw and all external systems. Use it to inspect OpenClaw agents, sessions, skills, tools, status, configuration, capabilities, and environment, and for actions such as messages, web search, lists, reminders, notes, research, drafts, scheduling, smart-home control, and app interactions. For every OpenClaw question or request, call this tool; never infer the answer from the camera. Speak exactly one short pending acknowledgement before calling. After the call, stop speaking and wait; never report success or a result until this tool returns. When in doubt, use this tool.",
    "parameters": [
      "type": "object",
      "properties": [
        "task": [
          "type": "string",
          "description": "Clear, detailed description of what to do. Include all relevant context: names, content, platforms, quantities, etc."
        ]
      ],
      "required": ["task"]
    ] as [String: Any]
  ]

  static let captureMedia: [String: Any] = [
    "name": "capture_media",
    "description": "Capture media from the active glasses session. Use snapshot for a fresh still image before answering a visual request that explicitly asks to take or save a picture. Video recording is not available through DAT 0.8; calling video returns an explicit native Meta fallback instead of pretending to record.",
    "parameters": [
      "type": "object",
      "properties": [
        "kind": [
          "type": "string",
          "enum": ["snapshot", "video"],
          "description": "The requested media type."
        ],
        "durationSeconds": [
          "type": "integer",
          "minimum": 1,
          "maximum": 30,
          "description": "Requested video duration. Ignored for snapshots."
        ]
      ],
      "required": ["kind"]
    ] as [String: Any]
  ]

  static func routeHarness(
    registry: NamedHarnessRegistry
  ) -> [String: Any] {
    [
      "name": "route_harness",
      "description": "Route a request to one registered named harness. Available names: \(registry.promptDescription). Use the name spoken by the user. Never silently substitute another harness; return the explicit fallback or unavailable status.",
      "parameters": [
        "type": "object",
        "properties": [
          "target": [
            "type": "string",
            "description": "Registered invocation name spoken by the user."
          ],
          "operation": [
            "type": "string",
            "enum": NamedHarnessOperation.allCases.map(\.rawValue),
            "description": "Scoped operation. Eva uses execute; Meta uses handoff; Codex uses only the listed task operations."
          ],
          "task": [
            "type": "string",
            "description": "The user's request or Codex continuation instruction."
          ],
          "taskReference": [
            "type": "string",
            "description": "Opaque task reference returned by the scoped Codex bridge."
          ],
          "actionReference": [
            "type": "string",
            "description": "Opaque action reference used only for operation_status and cancel_operation. Prepared continuations are confirmed exclusively in VisionClaw's trusted iPhone sheet."
          ],
          "clientRequestID": [
            "type": "string",
            "description": "Replay-safe request ID returned by prepare_continue. Reuse it unchanged only for that action's status or cancellation; continuation approval is available exclusively in VisionClaw's trusted iPhone sheet."
          ],
        ],
        "required": ["target", "operation"]
      ] as [String: Any]
    ]
  }
}
